### ISSUE TITLE
SDBC issue

### ISSUE THREAD

**Issue by plindsa** [gahansen/Albany#234]
I have been seeing some issues with the SDBC implementation in Albany.  Specificallly, it seems that the DBC columns are not being zeroed out properly.  This also causes the Jacobain for multi-processor runs to lose its symmetry.

I have tracked this down to a commit in late October by @ibaned  (d01d74c5f6af2493a6320ec29f318567f9939e09) where the SDBC code was refactored.  Running a build of Albany prior to the start of that refactor using the older version of the code by @ikalash appears to zero out the columns correctly.

I have attached two figures that show results for 2 and 4 processors.  Blue represents data locations from the serial run, white and black are the DBC rows/columns and diagonal positions respectively, and red are entries that are zeroed for serial runs but are missed when running in parallel.  From those figures, it can be seen that the parallel runs of Albany are not zeroing out all of the DBC columns.

![np2](https://user-images.githubusercontent.com/31855352/33783904-895550fc-dc13-11e7-8595-e645ccf89669.jpg)
![np4](https://user-images.githubusercontent.com/31855352/33783905-8971e604-dc13-11e7-971a-a6781fb94ac9.jpg)

**Comment by plindsa** [comment]
I should have mentioned this, but these results are from one of the tests in the repo... tests/small/LCM/TimeDependentSDBC/finer_cube_sdbc.yaml.

**Comment by ibaned** [comment]
@plindsa I think I've confirmed this, and fortunately I also think I have a fix.

**Comment by plindsa** [comment]
Awesome!

**Comment by ibaned** [comment]
@plindsa I just pushed commit db62c44f3, let me know if it helps.

**Comment by plindsa** [comment]
That fixed it - thanks!

**Comment by ikalash** [comment]
@ibaned , could you please propagate these changes into LCM::StrongSchwarzBC ?

**Comment by ibaned** [comment]
@ikalash the code duplication between these two worries me... can we separate out the common code somehow?

**Comment by ikalash** [comment]
I can look at whether they can be combined.  Perhaps @lxmota has thought about it already and had a reason not to combine them?  Alejandro, can you please comment?

**Comment by lxmota** [comment]
We can go ahead and combine them. That would be best.

**Comment by ikalash** [comment]
Per my phone discussion with @lxmota a few mins ago: it may not be entirely trivial to remove the code duplication in the Schwarz and regular Dirichlet BCs, according to @lxmota.  It seems to me it should be possible to call the regular Dirichlet BC (or parts of it) from the Schwarz BC, but we need to think about it and try to come up with a design.  I'll look at it and will probably open a separate issue on this at that time.
